gateway/eth_sub.go

```diff
@@ -0,0 +1,70 @@
+package gateway
+
+import (
+	"context"
+	"sync"
+
+	"github.com/filecoin-project/go-jsonrpc"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+)
+
+type EthSubHandler struct {
+	queued map[ethtypes.EthSubscriptionID][]ethtypes.EthSubscriptionResponse
+	sinks  map[ethtypes.EthSubscriptionID]func(context.Context, *ethtypes.EthSubscriptionResponse) error
+
+	lk sync.Mutex
+}
+
+func NewEthSubHandler() *EthSubHandler {
+	return &EthSubHandler{
+		queued: make(map[ethtypes.EthSubscriptionID][]ethtypes.EthSubscriptionResponse),
+		sinks:  make(map[ethtypes.EthSubscriptionID]func(context.Context, *ethtypes.EthSubscriptionResponse) error),
+	}
+}
+
+func (e *EthSubHandler) addSub(ctx context.Context, id ethtypes.EthSubscriptionID, sink func(context.Context, *ethtypes.EthSubscriptionResponse) error) error {
+	e.lk.Lock()
+	defer e.lk.Unlock()
+
+	for _, p := range e.queued[id] {
+		if err := sink(ctx, &p); err != nil {
+			return err
+		}
+	}
+	delete(e.queued, id)
+	e.sinks[id] = sink
+	return nil
+}
+
+func (e *EthSubHandler) removeSub(id ethtypes.EthSubscriptionID) {
+	e.lk.Lock()
+	defer e.lk.Unlock()
+
+	delete(e.sinks, id)
+	delete(e.queued, id)
+}
+
+func (e *EthSubHandler) EthSubscription(ctx context.Context, r jsonrpc.RawParams) error {
+	p, err := jsonrpc.DecodeParams[ethtypes.EthSubscriptionResponse](r)
+	if err != nil {
+		return err
+	}
+
+	e.lk.Lock()
+
```

Comment:
just defer unlock?

Reply:
No, we want to call the `sink` callback without the lock

Reply:
Ah, my bad!